### PR TITLE
Adding a configurable timeout for calls to the loki server.

### LINKED ISSUE #264: Add timeout to loki client in promtail
And make it clear in the error message whats happening.

### PR DESCRIPTION
**Pull request by slim-bean**
Also updating promtail test to set sane defaults for all configurations and then override for testing.
Removing .idea folder from .gitignore, we should encourage people (and myself) to use global ignore for IDE configs

Fixes #264

### CODE REVIEW

File: pkg/promtail/client/client.go [hunk 1]
```diff
@@ -155,8 +158,18 @@ func (c *Client) sendBatch(batch map[model.Fingerprint]*logproto.Stream) {
...
+		timedCtx, ctxCancel := context.WithTimeout(ctx, c.cfg.Timeout)
+		status, err = c.send(timedCtx, buf)
 		requestDuration.WithLabelValues(strconv.Itoa(status)).Observe(time.Since(start).Seconds())
+		ctxErr := timedCtx.Err()
+		if ctxErr != nil {
+			switch ctxErr {
+			case context.DeadlineExceeded:
+				err = errors.Wrap(err, "Timeout waiting for server")
+			}
+		}
```
Comment (tomwilkie) [inline pkg/promtail/client/client.go]: I'm not sure you need to check the context.Err here - the `client.Do()` in `c.send` will return what ever error comes from the context: https://golang.org/src/net/http/transport.go?s=3607:10104#L451

File: pkg/promtail/client/client.go [hunk 2]
```diff
@@ -155,8 +158,18 @@ func (c *Client) sendBatch(batch map[model.Fingerprint]*logproto.Stream) {
...
+		ctxCancel()
```
Comment (tomwilkie) [inline pkg/promtail/client/client.go]: We probably want to move the `context.WithTimeout` into `client.send` and then defer the `ctxCancel` call.  And just call it `cancel`.

File: pkg/promtail/promtail_test.go [hunk 3]
```diff
@@ -350,17 +350,23 @@ func buildTestConfig(t *testing.T, positionsFileName string, logDirName string)
...
-		BatchWait:      10 * time.Millisecond,
-		BatchSize:      10 * 1024,
-		ExternalLabels: nil,
+	cfg := config.Config{
+		ServerConfig:    server.Config{},
+		ClientConfig:    client.Config{},
+		PositionsConfig: positions.Config{},
+		ScrapeConfig:    []scrape.Config{},
+		TargetConfig:    targets.Config{},
 	}
```
Comment (tomwilkie) [inline pkg/promtail/promtail_test.go]: This is all a noop - they are all zero values.

Comment (slim-bean) [reply pkg/promtail/promtail_test.go]: Strange, I thought I needed those to call `flagext.RegisterFlags(&cfg)` but it appears I do not, I will remove them.

File: pkg/promtail/client/client.go (hunk 1 above)
Comment (slim-bean) [reply pkg/promtail/client/client.go]: the main reason for all this mess was to clarify the log message, without this you get a message that looks like 
```
level=warn msg="error sending batch, will retry" status=-1 error="Post http://10.10.10.10:3100/api/prom/push: context deadline exceeded"
```
So I added the check for the specific error to be able to add the "Timeout waiting for server" to the message.

I'll defer to you if you think the built in message is descriptive enough?

Comment (tomwilkie) [reply pkg/promtail/client/client.go]: `context deadline exceeded` is becoming a pretty common synonym for `timeout`, so yeah, I'd got for less code.

File: pkg/promtail/client/client.go [hunk 4]
```diff
@@ -191,11 +194,13 @@ func encodeBatch(batch map[model.Fingerprint]*logproto.Stream) ([]byte, error) {
 }
 
 func (c *Client) send(ctx context.Context, buf []byte) (int, error) {
+	timedCtx, cancel := context.WithTimeout(ctx, c.cfg.Timeout)
```
Comment (tomwilkie) [inline pkg/promtail/client/client.go]: This is a matter of opinion, but I'd overload the existing `ctx` here: 

```golang
ctx, cancel := context.WithTimeout(ctx, c.cfg.Timeout)
```

Its a pretty common pattern.

**Pull request merged**